### PR TITLE
Adding tracking to recommend the app feature.

### PR DESCRIPTION
**Pull request by develric**
Refs: pctCYC-8X-p2#comment-340 and  iOS [counter part PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/17027)

Part of #15269 this PR does the following:

- Adds tracking for  `recommend_app_engaged`: this event is fired when the Share Sheet is succesfully opened: it has a `source` parameter that currently is only triggered for "me" given the "about" source is not available
- Adds tracking for `recommend_app_content_fetch_failed`: this event is fired on failure in getting the remote template; it has a `source` parameter (only "me" available) and an `error` parameter that reports the error message. Currently the failure is triggered for
   - No network available
   - Volley network error
   - getTemplateFromJson parsing error
- A bit of refactor of calls provider to ease testability and moving the No network check inside the net calls provider so to have all failure causes in one place inside that class; also it seemed to me a good place where to check for network in the end.

### To test
- Activate the feature flag and the tracking in the app settings and look in logcat for events
- Check the following is triggered on success when opening the sheet (both when network call happens and in successive taps on the share button without leaving the Me screen)

```
🔵 Tracked: recommend_app_engaged, Properties: {"source":"me"}
```

- Check the`recommend_app_content_fetch_failed` event is triggered on following failures:
   - No Network available: setting the airplane mode you should get
```
🔵 Tracked: recommend_app_content_fetch_failed, Properties: {"source":"me","error":"getRecommendTemplate > No Network available"}
```
   - Volley network error, getTemplateFromJson parsing error: forcing conditions with debugger or by charles proxy (error property will depend on how the error is forced)
- Since a bit of refactoring was made, smoke test the feature.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing and unit testing added.

3. What automated tests I added (or what prevented me from doing so)
Unit testing for tracking added.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/src/main/java/org/wordpress/android/models/recommend/RecommendApiCallsProvider.kt [hunk 1]
```diff
@@ -8,89 +8,117 @@ import com.wordpress.rest.RestRequest.ErrorListener
...
+        } else {
+            val logMessage = "getTemplateFromJson > null response received"
+            logErrorAndTrack(source, logMessage)
+            Failure(contextProvider.getContext().getString(R.string.recommend_app_null_response))
+        }
     }
 
-    private fun getErrorStringAndLog(
-        functionName: String,
+    private fun getNetErroAndLogStrings(
```
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/models/recommend/RecommendApiCallsProvider.kt]: Small typo `Erro` → `Error` 

(that would be the correct spelling in Portuguese, though 😆)

File: WordPress/src/main/java/org/wordpress/android/ui/main/MeViewModel.kt [hunk 2]
```diff
@@ -135,10 +128,12 @@ class MeViewModel
             is ApiFetchedResult -> if (this.isError()) {
                 RecommendAppUiState(this.error!!)
             } else {
-                RecommendAppUiState(
+                val openSheetData = RecommendAppUiState(
                         link = this.link,
                         message = this.message
                 )
+                analyticsUtilsWrapper.trackRecommendAppEngaged(ME)
+                openSheetData
```
Comment (renanferrari) [inline WordPress/src/main/java/org/wordpress/android/ui/main/MeViewModel.kt]: Another way I've been handling this kind of thing is by using `.also`:

```suggestion
                RecommendAppUiState(
                        link = this.link,
                        message = this.message
                ).also {
                    analyticsUtilsWrapper.trackRecommendAppEngaged(ME)
                }
```

I think readability improves a bit, but it's totally a matter of preference, so feel free to ignore this.

**Review by renanferrari — COMMENTED** [review]
Hey @develric 👋 Thanks for the changes! I have tested the app and can confirm I see the correct messages on logcat both on success and failure. Code is looking pretty good as well! I just left two comments below with a typo I found and a suggestion for a very minor improvement. 

As we chatted on Slack, it seems that the `recommend_app_content_fetch_failed` event only shows up on Tracks if we remove the `error` property from it, so it seems that there's something that needs to be done there.

Other than that, everything looks good to go.

File: WordPress/src/main/java/org/wordpress/android/ui/main/MeViewModel.kt (hunk 2 above)
Comment (develric) [reply WordPress/src/main/java/org/wordpress/android/ui/main/MeViewModel.kt]: totally agree...and I continue to forget about `also` :)

**Review by renanferrari — APPROVED** [review]
LGTM! :shipit: 

Feel free to merge this once all checks are ok.

**Comment by develric** [comment]
Thanks @renanferrari for helping with the reviews of these PRs, much appreciated 🙇 ! Just leaving a note that as far as we got, to fix the `recommend_app_content_fetch_failed ` we needed to avoid using the `error` string as a key for tracking property. Changing it to `cause_of_issue` the event started to appear in tracks 🤷 ...cc @jkmassel in case it was a known issue or for any further insights 🙇

**Pull request merged**